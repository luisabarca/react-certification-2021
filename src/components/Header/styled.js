import styled from 'styled-components';

export const AppHeader = styled.header`
  box-shadow: 0px 2px 4px -1px rgb(0 0 0 / 20%);
  color: #fff;
  display: flex;
  min-height: 64px;
  width: 100%;
  z-index: 1100;
  box-sizing: border-box;
  flex-shrink: 0;
  flex-direction: column;
`;

export const SearchContainer = styled.div`
  background-color: rgba(255, 255, 255, 0.15);
  position: relative;
  border-radius: 5px;
  margin-right: 10px;
  margin-left: 0;
  width: 100%;
  display: flex;
  flex: 4;
`;

export const SearchIconContainer = styled.div`
  padding: 10px;
  height: 100%;
  position: absolute;
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const ToolbarSpacer = styled.div`
  display: flex;
  flex-grow: 10;
  flex: 10;
`;

export const ToolbarRight = styled.div`
  display: flex;
  flex: 5;
  justify-content: flex-end;
`;
