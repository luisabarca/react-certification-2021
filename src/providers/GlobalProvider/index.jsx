import React, { useContext, useState, useCallback } from 'react';
import { createTheme, CssBaseline, ThemeProvider } from '@material-ui/core';

export const GlobalContext = React.createContext(null);

export const useGlobalContext = () => {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error('GlobalContext not available');
  }

  return context;
};

export const GlobalProvider = ({ children }) => {
  const [query, setQuery] = useState('Wizeline');
  const [theme, setTheme] = useState('dark');

  const toggleTheme = useCallback(() => {
    setTheme(theme == 'light' ? 'dark' : 'light');
  }, [theme]);

  const selectedTheme = createTheme({
    palette: {
      type: theme,
    },
  });

  return (
    <GlobalContext.Provider value={{ query, setQuery, theme, setTheme, toggleTheme }}>
      <ThemeProvider theme={selectedTheme}>
      <CssBaseline />
        {children}
      </ThemeProvider>
    </GlobalContext.Provider>
  );
};

export default GlobalProvider;
