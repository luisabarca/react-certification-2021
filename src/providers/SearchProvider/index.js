import React, { useContext, useState } from 'react';

export const SearchContext = React.createContext(null);

export const useSearch = () => {
  const context = useContext(SearchContext);

  if (!context) {
    throw new Error('SearchProvider not available');
  }

  return context;
};

export const SearchProvider = ({ children }) => {
  const [query, setQuery] = useState('Wizeline');

  return (
    // eslint-disable-next-line react/jsx-filename-extension
    <SearchContext.Provider value={{ query, setQuery }}>
      {children}
    </SearchContext.Provider>
  );
};

export default SearchProvider;
