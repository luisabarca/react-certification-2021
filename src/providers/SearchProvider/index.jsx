import React, { useContext, useState } from 'react';

export const SearchContext = React.createContext(null);

export const useSearch = () => {
  const context = useContext(SearchContext);

  if (!context) {
    throw new Error('SearchProvider not available');
  }

  return context;
};

export const SearchProvider = ({ children }) => {
  const [query, setQuery] = useState('Wizeline');

  return (
    <SearchContext.Provider value={{ query, setQuery }}>
      {children}
    </SearchContext.Provider>
  );
};

export default SearchProvider;
