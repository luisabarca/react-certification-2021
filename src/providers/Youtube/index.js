import React, { useCallback, useContext, useEffect, useState } from 'react';

const YOUTUBE_API_KEY = process.env.REACT_APP_YOUTUBE_API_KEY;
const YoutubeApiContext = React.createContext(null);

export const useYoutubeApi = () => {
  const context = useContext(YoutubeApiContext);

  if (!context) {
    throw new Error('YoutubeApiProvider not available');
  }

  return context;
};

const loadClientScript = () => {
  const script = document.createElement('script');
  script.src = 'https://apis.google.com/js/client.js';

  return script;
};

export const YouTubeApiProvider = ({ children }) => {
  const { gapi } = window;
  const [ready, setReady] = useState(false);

  useEffect(() => {
    const script = loadClientScript();

    script.onload = () => {
      gapi.load('client', () => {
        // Set API KEY.
        gapi.client.setApiKey(YOUTUBE_API_KEY);
        gapi.client.load('youtube', 'v3', () => {
          setReady(true);
        });
      });
    };

    document.body.appendChild(script);
  }, [gapi]);

  const searchVideos = useCallback(async (query) => {
    const options = {
      part: ['snippet'],
      maxResults: 25,
      q: query,
    };

    try {
      const response = await window.gapi.client.youtube.search.list(options);
      return response.result;
    } catch (error) {
      return null;
    }
  }, []);

  const getVideo = useCallback(async (videoId) => {
    const options = {
      part: ['snippet,contentDetails,statistics'],
      id: videoId,
    };

    try {
      const response = await window.gapi.client.youtube.videos.list(options);
      return response.result;
    } catch (error) {
      return null;
    }
  }, []);

  if (!ready) return null;

  return (
    // eslint-disable-next-line react/jsx-filename-extension
    <YoutubeApiContext.Provider value={{ searchVideos, getVideo }}>
      {children}
    </YoutubeApiContext.Provider>
  );
};

export default YouTubeApiProvider;
