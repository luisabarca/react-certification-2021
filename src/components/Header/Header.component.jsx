import React, { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import {
  IconButton,
  InputBase,
  Toolbar,
  FormControlLabel,
  Switch,
  makeStyles,
  Typography,
  Menu,
  MenuItem,
  Drawer,
  List,
  ListItem,
} from '@material-ui/core';
import MenuIcon from '@material-ui/icons/Menu';
import SearchIcon from '@material-ui/icons/Search';
import { AccountCircle } from '@material-ui/icons';

import { useGlobalContext } from '../../providers/GlobalProvider';
import { useDebounce } from '../../utils/hooks/useDebounce';
import {
  SearchContainer,
  SearchIconContainer,
  AppHeader,
  ToolbarSpacer,
  ToolbarRight,
} from './styled';

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
  },
  inputInput: {
    color: 'white',
    padding: theme.spacing(1, 1, 1, 0),
    paddingLeft: `calc(1em + ${theme.spacing(4)}px)`,
    transition: theme.transitions.create('width'),
    width: '100%',
    [theme.breakpoints.up('md')]: {
      width: '20ch',
    },
  },
  toolbar: {
    width: '100%',
    display: 'flex',
    flex: 1,
  },
  title: {
    paddingRight: '15px',
  },
  drawer: {
    display: 'flex',
    width: 240,
  },
  appBar: {
    backgroundColor: theme.palette.type !== 'dark' ? '#1c5476' : '#556cd6',
  },
}));

const Header = () => {
  const classes = useStyles();
  const history = useHistory();
  const { query, setQuery, theme, toggleTheme, logout } = useGlobalContext();
  const [menuLeftEl, setMenuLeftEl] = useState(null);
  const [menuRightEl, setMenuRightEl] = useState(null);
  const [localQuery, setLocalQuery] = useState(query);

  const debounceValue = useDebounce(localQuery, 300);

  useEffect(() => {
    if (!debounceValue) return;

    setQuery(debounceValue);
  }, [debounceValue, setQuery]);

  const handleChange = (e) => {
    setLocalQuery(e.target.value);
  };

  const handleLeftMenu = (event) => {
    setMenuLeftEl(event.currentTarget);
  };

  const handleLeftMenuClose = () => {
    setMenuLeftEl(null);
  };

  const handleRightMenu = (event) => {
    setMenuRightEl(event.currentTarget);
  };

  const handleRightMenuClose = () => {
    setMenuRightEl(null);
  };

  const handleGoHome = (e) => {
    e.preventDefault();
    handleLeftMenuClose();
    history.push('/');
  };

  const handleGoLogin = (e) => {
    e.preventDefault();
    handleLeftMenuClose();
    history.push('/login');
  };

  const handleLogout = (e) => {
    e.preventDefault();
    handleLeftMenuClose();
    logout();
    history.push('/');
  };

  const renderLeftMenu = (
    <Drawer
      className={classes.drawer}
      anchor="left"
      open={Boolean(menuLeftEl)}
      onClose={handleLeftMenuClose}
      classes={{
        paper: classes.drawer,
      }}
    >
      <List>
        <ListItem button onClick={handleGoHome}>
          Home
        </ListItem>
        <ListItem button onClick={handleLogout}>
          Logout
        </ListItem>
      </List>
    </Drawer>
  );

  const renderRightMenu = (
    <Menu
      anchorEl={menuRightEl}
      anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
      id="header-right-menu"
      keepMounted
      transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      open={Boolean(menuRightEl)}
      onClose={handleRightMenuClose}
    >
      <MenuItem onClick={handleGoLogin}>Login</MenuItem>
      <MenuItem onClick={handleLogout}>Logout</MenuItem>
    </Menu>
  );

  return (
    <AppHeader className={classes.appBar}>
      <Toolbar className={classes.toolbar}>
        <IconButton
          edge="start"
          color="inherit"
          aria-label="open drawer"
          onClick={handleLeftMenu}
        >
          <MenuIcon />
        </IconButton>

        <Typography className={classes.title} variant="h6" noWrap>
          Wizeline Bootcamp
        </Typography>

        <SearchContainer>
          <SearchIconContainer>
            <SearchIcon />
          </SearchIconContainer>
          <InputBase
            placeholder="Search"
            value={localQuery}
            classes={{
              input: classes.inputInput,
            }}
            inputProps={{ 'aria-label': 'search' }}
            onChange={handleChange}
          />
        </SearchContainer>
        <ToolbarSpacer />
        <ToolbarRight>
          <FormControlLabel
            control={
              <Switch
                checked={theme === 'dark'}
                name="checkedB"
                color="primary"
                onChange={toggleTheme}
              />
            }
            label="Dark mode"
          />
          <IconButton aria-haspopup="true" onClick={handleRightMenu}>
            <AccountCircle fontSize="large" />
          </IconButton>
        </ToolbarRight>
      </Toolbar>
      {renderLeftMenu}
      {renderRightMenu}
    </AppHeader>
  );
};

export default Header;
