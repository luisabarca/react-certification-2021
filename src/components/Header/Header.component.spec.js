/* eslint-disable */
import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import Header from './Header.component';

import { SearchContext } from '../../providers/SearchProvider';

const renderHeaderWithSearchProvider = (snap) => {
  const query = 'Wizeline';
  const setQuery = jest.fn();
  const isSnap = snap || false;

  const component = (
    <SearchContext.Provider value={{ query, setQuery}}>
      <Header />
    </SearchContext.Provider>
  );

  if (isSnap) {
    return renderer.create(component);
  }

  return render(component);
};

describe('Header Component', () => {
  beforeEach(() => {
    renderHeaderWithSearchProvider();
  });

  it('should match snapshot', () => {
    const snap = renderHeaderWithSearchProvider(true);
  
    expect(snap.toJSON()).toMatchSnapshot();
  });

  it('should render a menu button', () => {
    const hamburgerButton = screen.getByRole('button', {
      name: 'open drawer',
    });

    expect(hamburgerButton).toBeInTheDocument();
  });

  it('should render a search input', () => {
    const component = screen.getByRole('textbox', {
      name: 'search',
    });

    expect(component).toBeInTheDocument();
  });

  it('should render a checkbox for theme switch', () => {
    const component = screen.getByRole('checkbox', {
      name: 'Dark mode',
    });

    expect(component).toBeInTheDocument();
  });
});
