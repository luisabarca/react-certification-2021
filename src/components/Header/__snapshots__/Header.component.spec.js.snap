// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component should match snapshot 1`] = `
<header
  className="sc-bdnxRM gNWeIq header"
>
  <div
    className="MuiToolbar-root MuiToolbar-regular makeStyles-toolbar-2 MuiToolbar-gutters"
  >
    <button
      aria-label="open drawer"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
      </span>
    </button>
    <div
      className="sc-gtsrHT irUJMy"
    >
      <div
        className="sc-dlnjwi iDvJWQ"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </div>
      <div
        className="MuiInputBase-root"
        onClick={[Function]}
      >
        <input
          aria-label="search"
          className="MuiInputBase-input makeStyles-inputInput-1"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Search"
          type="text"
          value="Wizeline"
        />
      </div>
    </div>
    <div
      className="sc-hKFxyN kqkEZC"
    />
    <div
      className="sc-eCApnc cpIVTH"
    >
      <label
        className="MuiFormControlLabel-root"
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-3 MuiSwitch-switchBase MuiSwitch-colorPrimary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-6 MuiSwitch-input"
                name="checkedB"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Dark mode
        </span>
      </label>
      <button
        aria-haspopup="true"
        className="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</header>
`;
